File: java/0844.0_Backspace_String_Compare.java
/*
stack
T: O(M + N)
S: O(M + N)

Runtime: 6 ms, faster than 13.13% of Java online submissions for Backspace String Compare.
Memory Usage: 42.6 MB, less than 23.96% of Java online submissions for Backspace String Compare.
*/
class Solution {
    public boolean backspaceCompare(String s, String t) {
        return build(s).equals(build(t));
    }
    
    public String build(String s) {
        Stack<Character> ans = new Stack();
        for (char c: s.toCharArray()) {
            if (c != '#')
                ans.push(c);
            else if (!ans.empty())
                ans.pop();
        }
        return String.valueOf(ans);
    }
}


/*
two pointers
T: O(M + N)
S: O(1)


Runtime: 1 ms, faster than 87.48% of Java online submissions for Backspace String Compare.
Memory Usage: 42.4 MB, less than 38.67% of Java online submissions for Backspace String Compare.
*/
class Solution {
    public boolean backspaceCompare(String s, String t) {
        int i = s.length() - 1, j = t.length() - 1;
        int skips = 0, skipt = 0;
        
        // while there may be chars in build(s) or build(t)
        while (i >= 0 || j >= 0) {
            // find position of next possible char in build(s)
            while (i >= 0) {
                if (s.charAt(i) == '#') {skips++; i--;}
                else if (skips > 0) {skips--; i--;}
                else break;
            }
            // find position of next possible char in build(t)
            while (j >= 0) {
                if (t.charAt(j) == '#') {skipt++; j--;}
                else if (skipt > 0) {skipt--; j--;}
                else break;
            }
            // if two actual characters are different
            if (i >= 0 && j >= 0 && s.charAt(i) != t.charAt(j)) {
                return false;
            }
            // if expecting to compare char vs nothing
            if ((i >= 0) != (j >= 0))
                return false;
            i--; j--;
        }
        return true;    
    }
}

File: cpp/0844.0_Backspace_String_Compare.cpp
/*
执行用时：4 ms, 在所有 C++ 提交中击败了9.73% 的用户
内存消耗：6 MB, 在所有 C++ 提交中击败了90.24% 的用户
通过测试用例：114 / 114
*/
class Solution {
private:
    void jump(const string &s, int &i, int &skip) {
        while (i >= 0) {
            if (s[i] == '#') skip++;
            else if (skip) skip--;
            else break;
            i--;
        }
    }

public:
    bool backspaceCompare(string s, string t) {
        int i = s.size() - 1, j = t.size() - 1;
        // backspace count
        int skips = 0, skipt = 0;
        while (i >= 0 || j >= 0){
            jump(s, i, skips);
            jump(t, j, skipt);
            if (i < 0 || j < 0) break;
            if (s[i--] != t[j--]) return false;
        }
        return i < 0 && j < 0;
    }
};

